cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/schedule/StreamSizeScheduler.java

```diff
@@ -122,4 +253,372 @@ private String getProgramScheduleId(Id.Program program, SchedulableProgramType p
+        return streamAdmin.fetchStreamSize(streamAdmin.getConfig(streamId));
+      } catch (IOException e) {
+        LOG.error("Could not poll size for stream {}", streamId);
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Wrapper around a {@link StreamSizeSchedule} which will run a program whenever it receives enough
+   * data from a stream, via notifications.
+   */
+  private final class StreamSizeScheduleTask {
+    private final Id.Program programId;
+    private final SchedulableProgramType programType;
+    private final StreamSizeSchedule streamSizeSchedule;
+
+    private long baseSize;
+    private long baseTs;
+    private boolean running;
+
+    private StreamSizeScheduleTask(Id.Program programId, SchedulableProgramType programType,
+                                   StreamSizeSchedule streamSizeSchedule) {
+      this.programId = programId;
+      this.programType = programType;
+      this.streamSizeSchedule = streamSizeSchedule;
+    }
+
+    public void startSchedule(long baseSize, long baseTs, boolean active, boolean persist) {
+      LOG.debug("Starting schedule {} with baseSize {}, baseTs {}, active {}. Should be persisted: {}",
+                streamSizeSchedule.getName(), baseSize, baseTs, active, persist);
+      this.baseSize = baseSize;
+      this.baseTs = baseTs;
+      this.running = active;
+    }
+
+    public boolean isRunning() {
+      return running;
+    }
+
+    public void received(StreamSizeNotification notification) {
+      if (!running) {
+        return;
+      }
+      long pastRunSize;
+      long pastRunTs;
+      synchronized (this) {
+        if (notification.getSize() < baseSize) {
+          // This can happen when a stream is truncated: the baseSize is still the old size,
+          // but we receive notification with way less data
+          baseSize = notification.getSize();
+          baseTs = notification.getTimestamp();
+          return;
+        }
+        if (notification.getSize() < baseSize + toBytes(streamSizeSchedule.getDataTriggerMB())) {
+          return;
+        }
+
+        // Update the baseSize as soon as possible to avoid races
+        pastRunSize = baseSize;
+        pastRunTs = baseTs;
+        baseSize = notification.getSize();
+        baseTs = notification.getTimestamp();
+        LOG.debug("Base size and ts updated to {}, {} for streamSizeSchedule {}",
+                  baseSize, baseTs, streamSizeSchedule);
+      }
+
+      Arguments args = new BasicArguments(ImmutableMap.of(
+        ProgramOptionConstants.SCHEDULE_NAME, streamSizeSchedule.getName(),
+        ProgramOptionConstants.LOGICAL_START_TIME, Long.toString(baseTs),
+        ProgramOptionConstants.RUN_DATA_SIZE, Long.toString(baseSize),
+        ProgramOptionConstants.PAST_RUN_LOGICAL_START_TIME, Long.toString(pastRunTs),
+        ProgramOptionConstants.PAST_RUN_DATA_SIZE, Long.toString(pastRunSize)
+      ));
+
+      while (true) {
```

Comment:
I think it would be better to retry only a few times. And if that fails, reset base time and size to what it was before (since they reflect the last run of the schedule).